components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/DefaultSessionManager.cs

```diff
@@ -226,20 +265,21 @@ public class DefaultSessionManager : ISessionManager {
         /// <inheritdoc/>
         public async Task RemoveSessionAsync(ConnectionModel connection, bool onlyIfEmpty = true) {
             var key = new ConnectionIdentifier(connection);
-            await _lock.WaitAsync();
             try {
+                await _lock.WaitAsync();
```

Comment:
Nit - i'd suggest doing the acquire outside and let it throw as that prevents the Release() from being called in that case, i.e. when wait did not capture the semaphore you do not want to release it (and end up with a bad semaphore count)

Reply:
right